src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -102,6 +102,132 @@ private PatternSyntax ParsePattern()
+                }
+
+                return _syntaxFactory.DeclarationExpression(type, designation);
+            }
+            else
+            {
+                return this.ParseSubExpression(Precedence.Expression);
+            }
+        }
+
+        private bool IsPossibleDeclarationExpression(ParseTypeMode mode, bool permitTupleDesignation)
+        {
+            if (this.IsInAsync && this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword)
+            {
+                // can't be a declaration expression.
+                return false;
+            }
+
+            var resetPoint = this.GetResetPoint();
+            try
+            {
+                SyntaxToken lastTokenOfType;
+                switch (ScanType(out lastTokenOfType))
+                {
+                    case ScanTypeFlags.PointerOrMultiplication:
+                        if (mode == ParseTypeMode.FirstElementOfPossibleTupleLiteral || mode == ParseTypeMode.AfterTupleComma)
+                        {
+                            // Tuples cannot contain pointer types because pointers may not be generic type arguments.
+                            return false;
+                        }
+                        break;
+
+                    case ScanTypeFlags.NotType:
+                        return false;
+                }
+
+                // check for a designation
+                if (!ScanDesignation(permitTupleDesignation && (lastTokenOfType.IsVar() || IsPredefinedType(lastTokenOfType.Kind))))
+                {
+                    return false;
+                }
+
+                return mode != ParseTypeMode.FirstElementOfPossibleTupleLiteral || this.CurrentToken.Kind == SyntaxKind.CommaToken;
+            }
+            finally
+            {
+                this.Reset(ref resetPoint);
+                this.Release(ref resetPoint);
+            }
+        }
+
+        private bool ScanDesignation(bool permitTuple)
+        {
+            switch (this.CurrentToken.Kind)
+            {
+                default:
+                    return false;
+                case SyntaxKind.IdentifierToken:
+                    var result = this.IsTrueIdentifier();
+                    this.EatToken();
+                    return result;
+                case SyntaxKind.OpenParenToken:
+                    if (!permitTuple)
+                    {
+                        return false;
+                    }
+
+                    bool sawComma = false;
+                    while (true)
+                    {
+                        this.EatToken(); // consume the `(` or `,`
+                        if (!ScanDesignation(true))
+                        {
+                            return false;
+                        }
+                        switch (this.CurrentToken.Kind)
+                        {
+                            case SyntaxKind.CloseParenToken:
+                                this.EatToken();
```

Comment:
As with other scan methods, it is unspecified what tokens have been eaten when it returns false.